legacy/signals/signals_test.go

```diff
@@ -0,0 +1,152 @@
+	sigTests := []sigTest{
+		{
+			signal: syscall.SIGIO,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGIO.String()),
+			},
+			terminate: false,
+		},
+		{
+			signal: syscall.SIGALRM,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGALRM.String()),
+			},
+			terminate: false,
+		},
+		{
+			signal: syscall.SIGALRM,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGALRM.String()),
+			},
+			terminate: false,
+		},
+		{
+			signal: syscall.SIGQUIT,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGQUIT.String()),
+				fmt.Sprintf("Exiting from signal: %s", syscall.SIGQUIT.String()),
+			},
+			terminate: true,
+		},
+		{
+			signal: syscall.SIGINT,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGINT.String()),
+				fmt.Sprintf("Exiting from signal: %s", syscall.SIGINT.String()),
+			},
+			terminate: true,
+		},
+		{
+			signal: syscall.SIGABRT,
+			expected: []string{
+				fmt.Sprintf("Encoutered signal: %s", syscall.SIGABRT.String()),
+				fmt.Sprintf("Exiting from signal: %s", syscall.SIGABRT.String()),
+			},
+			terminate: true,
+		},
+	}
+	// Capture all logs.
+	logs := []string{}
+	// Mock logrus.Debug statements.
+	signals.Debug = func(args ...interface{}) {
+		logs = append(logs, fmt.Sprint(args...))
+	}
+	// Determine if LogSignal invoked Stop().
+	terminated := func() bool {
+		return <-signals.ExitChannel
+	}
+	// Used for enforcing defaults before each test.
+	reset := func() {
+		logs = []string{}
+	}
+	// Run through all tests.
+	for _, st := range sigTests {
+		reset()
+		// Log the test signal.
+		signals.LogSignal(st.signal)
+		// Ensure the logs are correct.
+		require.EqualValues(t, st.expected, logs, "Unexpected signal observation logs.")
+		if st.terminate {
+			// Ensure Stop() was invoked if the test specifies.
+			require.True(t, terminated(), "LogSignal did not terminate on exit signal.")
+		}
+	}
+}
+
+func TestLogSignals(t *testing.T) {
+	// Capture concurrent function termination.
+	wg := sync.WaitGroup{}
+	terminated := false
+	// Ensure the test waits for logSignals to finish executing.
+	logSignals := func() {
+		signals.LogSignals()
+		terminated = true
+		wg.Done()
+	}
+	// Start logging.
+	wg.Add(1)
+	go logSignals()
+	// Pass multiple non-exit signals.
+	signals.SignalChannel <- syscall.SIGBUS
+	signals.SignalChannel <- syscall.SIGALRM
+	signals.SignalChannel <- syscall.SIGSYS
+	signals.SignalChannel <- syscall.SIGIO
+	// Force exit.
+	signals.Stop()
+	wg.Wait()
+	// Ensure termination happened.
+	require.True(t, terminated, "LogSignals() did not terminate on call to Stop()")
+
+	// Reset termination bool for new test.
+	terminated = false
+	// Start logging.
+	wg.Add(1)
+	go logSignals()
+	// Pass multiple non-exit signals.
+	signals.SignalChannel <- syscall.SIGTTOU
+	signals.SignalChannel <- syscall.SIGCHLD
+	signals.SignalChannel <- syscall.SIGPIPE
```

Comment:
Nit: Unless you're performing some sort of checking of state in between these non-exit signals, these insertions into the channel are questionable. "If you cannot measure it, it is meaningless."

Reply:
SignalChannel is buffered for 1 signal at a time. Therefore, if the consumer (LogSignals) does not consume them, we will block and the test will fail. This is why I send multiple signals, ensuring we are actually handling them. I'll mention this in a comment.